[![](https://img.shields.io/badge/Pypi-v1.2.3-519dd9.svg)](https://pypi.org/project/MOFF/)
[![License: GUN](https://img.shields.io/badge/License-GUN-yellow.svg)](https://github.com/MDhewei/MOFF/blob/master/LICENSE)
![](https://img.shields.io/badge/language-python-orange.svg)

## Introduction of MOFF 
  
                                # # # # #   # # # #   # # # #  # # # # 
                                #   #   #   #     #   #        #       
                                #   #   #   #     #   # # # #  # # # #    
                                #   #   #   #     #   #        #          
                                #   #   #   # # # #   #        #          
                                     

    Hi, my name is MOFF. I was designed to predict off-target effects for CRISPR/Cas9! 
    
    I have three major functions: 1). Predict off-target effects for any given gRNA-target pair.
                                  2). Predict genome-wide off-target effects for any sgRNA.
                                  3). Select best sgRNAs for allele-specific knockout.
    
    
    Hope you enjoy playing with me ^o^!
                                     
    Any questions or bugs, please contact hwkobe.1027@gmail.com or whe3@mdanderson.org
                                     


## How to install MOFF


### Requirements

- Python>=3.4
- pandas, numpy, scipy
- tensorflow, keras
 
### Installation 
 Step1: Install Anaconda (highly recomended)
    
 ```console
 wget https://repo.continuum.io/archive/Anaconda2-2018.12-Linux-x86_64.sh 
 bash Anaconda2-2018.12-Linux-x86_64.sh 
 ```

 Step2: Install MOFF through pip
 ```console     
 pip install MOFF
 ```
    
 Step3: **OR** you can install MOFF through git clone
 ```console   
 git clone https://github.com/MDhewei/MOFF.git
 cd MOFF
 python setup.py install
 ```
    

## How to use MOFF

### 1. MOFF score: Predict off-target effects for given gRNA-target pairs


#### Inputs for MOFF score

MOFFscore requires the user to provide .csv or .txt file containing sgRNA sequences and corresponding 
DNA target sequences. Each line should have one gRNA(20bp+PAM) and one target(20bp+PAM) sequence. Note that MOFF is designed
for mismatch-only off-target prediction, not for indel mutations. See example file [here.](https://github.com/MDhewei/MOFF/blob/master/MOFF/MOFFscore_test.txt)

                    
     Arguments of the program:

     -i/--inputfile (required): 
     Input files containing sgRNA sequences and corresponding DNA target sequences.
 
     -p/--prefix (Optional): 
     Prefix of the file to save the outputs, default: ScoreTest.

     -o/--outputdir (Optional): 
     Directory to save output files, if no directory is given a folder named MOFF_scores
     will be generated in current working directory.
 

#### Example to run MOFFscore

```console
MOFF score -i MOFFscore_test.txt
```

#### Columns of Output table

    - crRNA: the sgRNAs designed to target specific DNA sequence
    - DNA: the DNA sequence of the off-target site 
    - MDE: predicted mismatch-dependent off-target effect 
    - CE:  predicted combinatorial effect
    - MMs: the number of mismatches between sgRNA and off-target
    - GMT: predicted guide-intrinsic mismatch tolerence 
    - MOFF: the final MOFF score predicted for given gRNA-target pair



### 2. MOFF aggregate: Predict the genome-wide off-target effects for given sgRNAs

  MOFF aggregation can directly take the outputs of [CRISPRitz](https://github.com/pinellolab/CRISPRitz) and [Cas-OFFinder](http://www.rgenome.net/cas-offinder/) as inputs. Output table files generated by any genome-wide off-target searching methods are also supported in theory, but the columns of outputs for different methods are different, therefore, it is required to modify the column name of sgRNA(20bp+PAM) and target(20bp+PAM) to 'crRNA' and 'DNA' respectively. Note that MOFF only supports mismatch-only off-target predictions, indel mutations are not applicable. File formats including .csv and .txt are accepted. See examplefile [here.](https://github.com/MDhewei/MOFF/blob/master/MOFF/MOFFaggregation_test.txt)

     Arguments of the program:

     -i/--inputfile (required): 
     Input files containing all the potential off-target sites in the genome for given sgRNA(s).

     -p/--prefix (Optional): 
     Prefix of the file to save the outputs, default: AggregationTest.

     -o/--outputdir (Optional): 
     Directory to save output files, if no directory is given a folder named MOFF_aggregation
     will be generated in current working directory.
 

#### Example to run MOFF aggregate

```console
MOFF aggregate -i MOFFaggregation_test.txt
```

#### Columns of Output table

    - sgRNA: the sgRNAs selected to predict genome-wide off-target
    - GMT: aggregated guide-intrinsic mismatch tolerence 
    - MDE.sum: aggregated mismatch-dependent off-target effect 
    - MOFF.sum: aggregated MOFF score for specific sgRNA
